src/emulator/kernel/src/load_self.cpp

```diff
@@ -0,0 +1,335 @@
+#include <kernel/types.h>
+
+#include <nids/functions.h>
+#include <util/log.h>
+
+#include <elfio/elf_types.hpp>
+#define SCE_ELF_DEFS_TARGET
+#include <sce-elf-defs.h>
+#undef SCE_ELF_DEFS_TARGET
+#include <self.h>
+
+#define NID_MODULE_STOP     0x79F8E492
+#define NID_MODULE_EXIT     0x913482A9
+#define NID_MODULE_START    0x935CD196
+#define NID_MODULE_INFO     0x6C2224BA
+#define NID_PROCESS_PARAM   0x70FBA1E7
+
+#include <miniz.h>
+
+#include <assert.h>
+#include <cstring>
+#include <iomanip>
+#include <iostream>
+
+using namespace ELFIO;
+
+static const bool LOG_IMPORTS = false;
+static const bool LOG_EXPORTS = false;
+
+static bool load_var_imports(const uint32_t *nids, const Ptr<uint32_t> *entries, size_t count, const MemState &mem) {
+    for (size_t i = 0; i < count; ++i) {
+        const uint32_t nid = nids[i];
+        const Ptr<uint32_t> entry = entries[i];
+        
+        if (LOG_IMPORTS) {
+            const char *const name = import_name(nid);
+            LOG_DEBUG("\tNID {:#08x} ({}) at {:#x}", nid, name, entry.address());
+        }
+        
+        /*uint32_t *const stub = entry.get(mem);
+        stub[0] = 0xef000000; // svc #0 - Call our interrupt hook.
+        stub[1] = 0xe1a0f00e; // mov pc, lr - Return to the caller.
+        stub[2] = nid; // Our interrupt hook will read this.*/
+    }
+    
+    return true;
+}
+
+static bool load_func_imports(const uint32_t *nids, const Ptr<uint32_t> *entries, size_t count, const MemState &mem) {
+    for (size_t i = 0; i < count; ++i) {
+        const uint32_t nid = nids[i];
+        const Ptr<uint32_t> entry = entries[i];
+
+        if (LOG_IMPORTS) {
+            const char *const name = import_name(nid);
+            LOG_DEBUG("\tNID {:#08x} ({}) at {:#x}", nid, name, entry.address());
+        }
+
+        uint32_t *const stub = entry.get(mem);
+        stub[0] = 0xef000000; // svc #0 - Call our interrupt hook.
+        stub[1] = 0xe1a0f00e; // mov pc, lr - Return to the caller.
+        stub[2] = nid; // Our interrupt hook will read this.
+    }
+
+    return true;
+}
+
+static bool load_imports(const sce_module_info_raw &module, Ptr<const void> segment_address, const MemState &mem) {
+    const uint8_t *const base = segment_address.cast<const uint8_t>().get(mem);
+    const sce_module_imports_raw *const imports_begin = reinterpret_cast<const sce_module_imports_raw *>(base + module.import_top);
+    const sce_module_imports_raw *const imports_end = reinterpret_cast<const sce_module_imports_raw *>(base + module.import_end);
+
+    for (const sce_module_imports_raw *imports = imports_begin; imports < imports_end; imports = reinterpret_cast<const sce_module_imports_raw *>(reinterpret_cast<const uint8_t *>(imports) + imports->size)) {
+        if (LOG_IMPORTS) {
+            const char *const lib_name = Ptr<const char>(imports->module_name).get(mem);
+            LOG_INFO("Loading imports from {}", lib_name);
+        }
+
+        //assert(imports->version == 1);
```

Comment:
why was this commented? please don't comment asserts if not completely sure, and when you do just remove the line